staging/src/k8s.io/externaljwt/pkg/plugin/plugin.go

```diff
@@ -0,0 +1,263 @@
+func (p *Plugin) GenerateToken(claims *jwt.Claims, privateClaims interface{}) (string, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
+	defer cancel()
+
+	claims.Issuer = p.iss
+
+	mergedClaims, err := mergeClaims(claims, privateClaims)
+	if err != nil {
+		return "", fmt.Errorf("while merging claims: %w", err)
+	}
+
+	payload := base64.RawURLEncoding.EncodeToString(mergedClaims)
+
+	request := &ejspb.SignJWTRequest{
+		Claims: payload,
+	}
+
+	response, err := p.client.Sign(ctx, request)
+	if err != nil {
+		return "", fmt.Errorf("while signing jwt: %w", err)
+	}
+
+	if err := p.validateJWTHeader(response); err != nil {
+		return "", fmt.Errorf("while validating header: %w", err)
+	}
+
+	return response.GetHeader() + "." + payload + "." + response.GetSignature(), nil
+}
+
+// GetTokenVerificationKeys returns a map of supported external keyIDs to keys
+// the keys are PKIX-serialized.
+func (p *Plugin) GetTokenVerificationKeys(ctx context.Context) (*VerificationKeys, int, error) {
+	ctx, cancel := context.WithTimeout(ctx, 60*time.Second)
+	defer cancel()
+
+	req := &ejspb.FetchKeysRequest{}
+	resp, err := p.client.FetchKeys(ctx, req)
+	if err != nil {
+		return nil, 0, fmt.Errorf("while getting externally supported jwt signing keys: %w", err)
+	}
+
+	keys := []serviceaccount.PublicKey{}
+	for _, protoKey := range resp.GetKeys() {
+		parsedPublicKey, err := x509.ParsePKIXPublicKey(protoKey.GetKey())
+		if err != nil {
+			return nil, 0, fmt.Errorf("while parsing external public keys: %w", err)
+		}
+
+		keys = append(keys, serviceaccount.PublicKey{
+			KeyID:                    protoKey.GetKeyId(),
+			PublicKey:                parsedPublicKey,
+			ExcludeFromOidcDiscovery: protoKey.ExcludeFromOidcDiscovery,
+		})
+	}
+
+	vk := &VerificationKeys{
+		Keys:          keys,
+		DataTimestamp: resp.GetDataTimestamp().AsTime(),
+	}
+
+	return vk, int(resp.RefreshHintSeconds), nil
+}
+
+// GetServiceMetadata returns metadata associated with externalJWTSigner
+// It Includes details like max token lifetime supported by externalJWTSigner, etc.
+func (p *Plugin) GetServiceMetadata(ctx context.Context) (*Metadata, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
+	defer cancel()
+
+	req := &ejspb.MetadataRequest{}
+	resp, err := p.client.Metadata(ctx, req)
+	if err != nil {
+		return nil, fmt.Errorf("while getting metadata from external jwt signer: %w", err)
+	}
+
+	return &Metadata{
+		MaxTokenExpirationSeconds: int(resp.MaxTokenExpirationSeconds),
+	}, nil
+}
+
+// IsExternal informs if TokenGenerator intigerates with external-jwt-signer.
+func (p *Plugin) IsExternal() bool {
+	return true
+}
+
+// setKeyCache is for use with testcases.
+// It is required because creating keyCache itself requires a reference to Plugin.
+func (p *Plugin) setKeyCache(keyCache *KeyCache) {
+	p.keyCache = keyCache
+}
+
+func mergeClaims(claims *jwt.Claims, privateClaims interface{}) ([]byte, error) {
```

Comment:
make this a method both our static and external signer can use to merge claims

```
// MergeClaims returns a merged set of claims.
// Precedence of overlapping claims is privateClaims < claims < issuer
func MergeClaims(privateClaims interface{}, claims *jwt.Claims, issuer string) (map[string]interface{}, error)
```

Reply:
Done